### ISSUE TITLE
Swapping files in the NABU directory

### ISSUE THREAD

**Issue by nthieberger** [nabu-catalog/nabu#623]
Jono, is it possible for us to swap files directly into the file system in NABU? This would allow us to keep DOIs for files that we are now replacing due to length discrepancies. The problem with doing this up to now is that indexing fails to recognise the new file and it is not then accessible in NABU.

**Comment by JonoGillett** [comment]
@nthieberger if you replace the file in Nabu by uploading a new file with the same name then it should keep the same DOI (since the DOI is attached to the essence, not the filesystem file).

Can you elaborate on "the new file is not then accessible"? Since the import process simply copies the file over the top of the existing catalog version it should be accessible via the original essence file in the catalog.

**Comment by nthieberger** [comment]
I don't think the file is available in Nabu if I just sftp it into the
collection. Nabu doesn't seem to recognise it.

**Comment by JonoGillett** [comment]
@nthieberger Ah, creating/updating an essence file requires going through the import process. However as I said you should be able to use the same filename to update the existing record (and keep the same DOI).

You are correct that Nabu does not detect (or periodically check) changes to files directly on the server e.g. via SFTP.

**Comment by nthieberger** [comment]
Yes, so that's to be tested. I suspect that putting a new file through the
normal ingest will create a new DOI. If the item is private then it will
generate a new DOI when the private status is changed. So is there some way
to suppress that happening ?

**Comment by JonoGillett** [comment]
@nthieberger just to confirm, when you say "new file" are you saying something like "ABC-123-A_fixed.mp3" ? or simply putting a fixed version of "ABC-123-A.mp3" in the send to Nabu directory?

If you find that there is a need to suppress DOI generation for certain cases then we can investigate that, but I believe we should be able to work within the existing processes in this instance.

**Comment by AmandaPDSC** [comment]
By new file, we mean putting a new version of the file (with the same filename) - so in your example - ABC-123-A.mp3 - to overwrite the previous version

**Comment by JonoGillett** [comment]
If that is the case then the system will update the Essence metadata (size, duration, bitrate etc) but the DOI attached to the Essence would remain the same.
Currently the DOI minting fetches public records which don't already have DOIs and then creates them, so anything that already has one will remain unchanged.

**Comment by jodiekell** [comment]
So we will try this with a file that needs to be replaced by checking the DOI, replacing the file through the normal Azoulay ingestion, and then checking the DOI once it is replaced. It is from the GS1 collection that is not private.

**Comment by jodiekell** [comment]
Using GS1-H3a as a test case, we have found that replacing the problematic audio files with fixed files (with the same name) through the usual ingestion process does not change the DOI, noting that this is not a file that is marked as private. 
So as Jono commented above, the system updated the Essence metadata but the DOI attached to the Essence remained the same.

**Comment by nthieberger** [comment]
Have you tried finding that file in NABU and playing it and downloading it? It could be that it is not recognised by NABU

**Comment by jodiekell** [comment]
Yes I have checked both files in NABU, they play and download correctly. The file length discrepancy is solved.

**Comment by nthieberger** [comment]
Excellent!

**Comment by caratgold** [comment]
Yay

**Comment by juliamiller** [comment]
Hey, that's cool! So I have loads of .wav files here that I can transfer up at some point to replace the truncated versions in Nabu. How should I go about doing this? Do I just send them through the same process? Or do I get them to you by other means and you manually put them in the archive?

**Comment by nthieberger** [comment]
Sounds like it is the normal process of ingestion

**Comment by jodiekell** [comment]
We think that the problem with the GS1 audio files was that they were peaking in volume (too hot) and at the time they were ingested this wasn't dealt with in the Dobbin workflow. This would have meant that the WAV files were creating problems as they were ingested into NABU and thus perhaps were truncated due to some kind of glitch. However since then Nick FG has implemented a leveler in the Dobbin workflow and so even if this is the case it will not be a problem.
With the GS1 files I am checking them in Wavelab and cropping them etc as we do for any files and then sending them through the normal ingestion process.
So Julia, you should be able to put the files through the normal ingestion process and we can check the file lengths as they go through. What collections do you have?

**Comment by juliamiller** [comment]
The problem with GS1 does not explain all the files I sent through, as those were not "too hot". But things lately have been great, so yay.

I have most files I have sent through. Ever. So I can send some in groups. I can start with some of the WSDS1 files. The long file length discrepancy spreadsheet can be found on Trello (https://trello.com/c/6mXPMiry); I had highlighted the files I have sent through.

**Comment by jodiekell** [comment]
Thanks Julia.
It seems that the discrepancies have been caused by a range of factors (one of which is the levels). We are isolating causes and checking that it is not continuing. At the moment it looks like we have solved any further truncating.
The main priorities now are replacing problematic files and we are looking at setting up a checking system to make sure this does not occur again (as in if there is future problems that cause file length discrepancies we will be notified and can act upon it quickly).
I did have a question, and that is how you generated the file length discrepancy spreadsheet that is found on trello. It would be good to generate another one once we have replaced all the files we have access to.
Also I have put some information into the trello card regarding the process we are enacting regarding this situation.
